File: golang/agents/inprocess_test.go
// Tests for ADR 0030 / issue #95: agents.Options.InProcess, the semantic
// counterpart to Options.Transport, so a host whose model is a Go function
// doesn't have to hand-build an http.RoundTripper to reach the sub-agent
// runtime. Ported from the spike at spikes/inprocess-subagent/SPIKE.md.
package agents

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	tn "github.com/muthuishere/toolnexus/golang"
)

// scriptedModel is the fake model shared by BOTH the top-level client and the
// sub-agent runtime below — the reporter's actual case. It never touches
// HTTP: it is handed the assembled InProcessRequest and returns one
// InProcessResponse. It instruments concurrency so the gate test can assert
// on real overlap, not just a trust-me counter.
type scriptedModel struct {
	inFlight int32
	maxSeen  int32
	overlaps int32 // count of calls that observed >1 in flight
	calls    int32
	hold     time.Duration
}

func (m *scriptedModel) generate(req tn.InProcessRequest) (tn.InProcessResponse, error) {
	n := atomic.AddInt32(&m.inFlight, 1)
	defer atomic.AddInt32(&m.inFlight, -1)
	atomic.AddInt32(&m.calls, 1)
	for {
		cur := atomic.LoadInt32(&m.maxSeen)
		if n <= cur || atomic.CompareAndSwapInt32(&m.maxSeen, cur, n) {
			break
		}
	}
	if n > 1 {
		atomic.AddInt32(&m.overlaps, 1)
	}
	if m.hold > 0 {
		time.Sleep(m.hold) // widen the window so a real bypass WOULD overlap
	}
	return tn.InProcessResponse{Content: fmt.Sprintf("ok model=%s turn-done", req.Model)}, nil
}

// TestInProcessOption_SharedGenerate_TopLevelAndSubAgent proves the reporter's
// actual case: ONE generate function serving BOTH a top-level
// tn.CreateInProcessClient AND a sub-agent runtime (via the new
// Options.InProcess field), with no copied adapter code — NewRuntime builds
// its transport by calling the SAME tn.InProcessTransport export that
// CreateInProcessClient calls.
func TestInProcessOption_SharedGenerate_TopLevelAndSubAgent(t *testing.T) {
	m := &scriptedModel{}

	// Top-level client, driven by m.generate directly.
	client := tn.CreateInProcessClient(tn.InProcessOptions{Model: "spike-model", Generate: m.generate})
	toolkit, err := tn.CreateToolkit(context.Background(), tn.Options{})
	if err != nil {
		t.Fatal(err)
	}
	topRes, err := client.Ask(context.Background(), "hello", toolkit, "top-level")
	if err != nil {
		t.Fatal(err)
	}
	if topRes.Status != "done" {
		t.Fatalf("top-level status = %q, want done (%s)", topRes.Status, topRes.Text)
	}

	// Sub-agent runtime, driven by the SAME m.generate via Options.InProcess.
	rt := NewRuntime(Options{
		InProcess: m.generate,
		Registry: map[string]Def{
			"worker": {Name: "worker", Does: "a scripted worker"},
		},
	})
	h, err := rt.Spawn(rt.Root, "worker", nil)
	if err != nil {
		t.Fatal(err)
	}
	rt.Wake(h, "do the thing")
	subRes := rt.Wait(h, 0)
	if subRes.Status != "done" {
		t.Fatalf("sub-agent status = %q, want done (%s)", subRes.Status, subRes.Text)
	}

	if got := atomic.LoadInt32(&m.calls); got != 2 {
		t.Fatalf("m.calls = %d, want 2 (one top-level Ask, one sub-agent turn)", got)
	}
}

// TestNewRuntime_InProcessAndTransportConflict proves construction-time
// validation, never precedence: setting both Options.Transport and
// Options.InProcess must error loudly (a panic, matching this constructor
// family's existing style — see tn.CreateInProcessClient).
func TestNewRuntime_InProcessAndTransportConflict(t *testing.T) {
	m := &scriptedModel{}
	defer func() {
		if recover() == nil {
			t.Fatal("NewRuntime did not panic with both Transport and InProcess set")
		}
	}()
	NewRuntime(Options{
		Transport: tn.InProcessTransport(m.generate),
		InProcess: m.generate,
	})
}

// TestNewRuntime_InProcessAndLLMConflict is the same validation for the
// InProcess/LLM pair.
func TestNewRuntime_InProcessAndLLMConflict(t *testing.T) {
	m := &scriptedModel{}
	defer func() {
		if recover() == nil {
			t.Fatal("NewRuntime did not panic with both LLM and InProcess set")
		}
	}()
	NewRuntime(Options{
		LLM:       &LLMOptions{BaseURL: "http://example.invalid"},
		InProcess: m.generate,
	})
}

// TestInProcessOption_GateHoldsAtConcurrencyOne is the falsifiable gate test
// (ADR 0030 gate item 2): MaxConcurrentTurns=1, five workers woken
// concurrently on the NEW Options.InProcess path. If the global turn gate
// (gatedTransport, runtime.go) stopped wrapping the transport built from
// Options.InProcess, scriptedModel would observe >1 in flight and this test
// fails.
func TestInProcessOption_GateHoldsAtConcurrencyOne(t *testing.T) {
	m := &scriptedModel{hold: 15 * time.Millisecond}
	rt := NewRuntime(Options{
		InProcess:          m.generate,
		MaxConcurrentTurns: 1,
		Registry: map[string]Def{
			"worker": {Name: "worker", Does: "a scripted worker"},
		},
	})
	const n = 5
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			h, err := rt.Spawn(rt.Root, "worker", nil)
			if err != nil {
				t.Error(err)
				return
			}
			rt.Wake(h, fmt.Sprintf("job %d", i))
			r := rt.Wait(h, 0)
			if r.Status != "done" {
				t.Errorf("worker %d status = %q, want done", i, r.Status)
			}
		}(i)
	}
	wg.Wait()

	if got := atomic.LoadInt32(&m.overlaps); got != 0 {
		t.Fatalf("gate FALSIFIED: %d call(s) observed >1 in flight (maxSeen=%d) "+
			"with MaxConcurrentTurns=1 — the global turn gate did not wrap the "+
			"Options.InProcess transport", got, m.maxSeen)
	}
	if got := rt.MaxObservedConcurrentTurns(); got != 1 {
		t.Fatalf("rt.MaxObservedConcurrentTurns() = %d, want 1", got)
	}
	t.Logf("gate holds: calls=%d maxSeen=%d overlaps=%d rt.MaxObservedConcurrentTurns=%d",
		m.calls, m.maxSeen, m.overlaps, rt.MaxObservedConcurrentTurns())
}

// TestInProcessOption_GateControl_WouldCatchABypass is the negative control:
// same 5 concurrent workers, but MaxConcurrentTurns=5 (effectively no gate).
// Proves the assertion above is a real detector, not a tautology — with room
// to run concurrently, scriptedModel DOES observe overlap.
func TestInProcessOption_GateControl_WouldCatchABypass(t *testing.T) {
	m := &scriptedModel{hold: 15 * time.Millisecond}
	rt := NewRuntime(Options{
		InProcess:          m.generate,
		MaxConcurrentTurns: 5,
		Registry: map[string]Def{
			"worker": {Name: "worker", Does: "a scripted worker"},
		},
	})
	const n = 5
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			h, err := rt.Spawn(rt.Root, "worker", nil)
			if err != nil {
				t.Error(err)
				return
			}
			rt.Wake(h, fmt.Sprintf("job %d", i))
			rt.Wait(h, 0)
		}(i)
	}
	wg.Wait()

	if got := atomic.LoadInt32(&m.overlaps); got == 0 {
		t.Fatalf("control INVALID: expected overlap with MaxConcurrentTurns=5 and a "+
			"15ms hold, got 0 overlaps (maxSeen=%d) — the detector proves nothing", m.maxSeen)
	}
	t.Logf("control confirms detector is live: calls=%d maxSeen=%d overlaps=%d",
		m.calls, m.maxSeen, m.overlaps)
}

File: elixir/test/agents/inprocess_test.exs
defmodule Toolnexus.Agents.InProcessTest do
  @moduledoc """
  ADR 0030 / issue #95: `Toolnexus.Agents.Runtime`'s `:in_process` option, the
  semantic counterpart to `:transport`, so a host whose model is an Elixir function
  doesn't have to hand-build a `:transport` to reach the sub-agent runtime. Ported
  from the Go reference (`golang/agents/inprocess_test.go`) and the spike at
  `spikes/inprocess-subagent/SPIKE.md`.
  """
  use ExUnit.Case, async: false

  alias Toolnexus.Client
  alias Toolnexus.Agents.Runtime

  @registry %{
    "worker" => %{name: "worker", description: "a scripted worker", system_prompt: "", model: "m-worker"}
  }

  # ---- scripted model, shared by BOTH the top-level client and the sub-agent
  # runtime below — the reporter's actual case. It never touches HTTP: it is
  # handed the assembled in-process request (%{messages, tools, model, body}) and
  # returns one assistant message (%{content: ...}). It instruments concurrency
  # (via an Agent-held counter set) so the gate tests assert on real overlap, not
  # just a trust-me counter.

  defp start_model(hold_ms \\ 0) do
    {:ok, pid} = Agent.start_link(fn -> %{in_flight: 0, max_seen: 0, overlaps: 0, calls: 0} end)
    %{pid: pid, hold_ms: hold_ms}
  end

  defp model_generate(%{pid: pid, hold_ms: hold_ms}, req) do
    n =
      Agent.get_and_update(pid, fn st ->
        n = st.in_flight + 1
        overlaps = if n > 1, do: st.overlaps + 1, else: st.overlaps
        st = %{st | in_flight: n, calls: st.calls + 1, max_seen: max(st.max_seen, n), overlaps: overlaps}
        {n, st}
      end)

    if hold_ms > 0, do: Process.sleep(hold_ms)
    Agent.update(pid, fn st -> %{st | in_flight: st.in_flight - 1} end)
    %{content: "ok model=#{req.model} turn-done, in-flight was #{n}"}
  end

  defp model_stats(%{pid: pid}), do: Agent.get(pid, & &1)

  defp bare_toolkit do
    {:ok, tk} = Toolnexus.create_toolkit(builtins: false)
    tk
  end

  test "one generate function serves BOTH the top-level in-process client and a sub-agent runtime, with no copied adapter code" do
    m = start_model()
    generate = fn req -> model_generate(m, req) end

    # Top-level client, driven by `generate` directly.
    client = Client.create_in_process(model: "spike-model", generate: generate)
    top = Client.run(client, "hello", bare_toolkit())
    assert top.status == "done"

    # Sub-agent runtime, driven by the SAME `generate` via Runtime's `:in_process`
    # option — which internally calls the SAME Toolnexus.Client.in_process_transport/1
    # export create_in_process/1 used above. No second adapter.
    rt = Runtime.new(%{in_process: generate, registry: @registry})
    {:ok, worker} = Runtime.spawn_agent(rt, Runtime.root(rt), "worker")
    Runtime.wake(rt, worker, "do the thing")
    sub = Runtime.wait(rt, worker, 5_000)
    assert sub.status == "done"

    assert model_stats(m).calls == 2, "one top-level Ask + one sub-agent turn"

    Runtime.shutdown(rt)
  end

  test "construction-time validation: :transport and :in_process together raise ArgumentError, never resolved by precedence" do
    m = start_model()
    generate = fn req -> model_generate(m, req) end
    transport = Client.in_process_transport(generate)

    assert_raise ArgumentError, ~r/mutually exclusive/, fn ->
      Runtime.new(%{transport: transport, in_process: generate, registry: @registry})
    end
  end

  test "construction-time validation: :llm and :in_process together raise ArgumentError" do
    m = start_model()
    generate = fn req -> model_generate(m, req) end

    assert_raise ArgumentError, ~r/mutually exclusive/, fn ->
      Runtime.new(%{llm: %{base_url: "http://example.invalid"}, in_process: generate, registry: @registry})
    end
  end

  test "the global turn gate holds at concurrency 1 on the :in_process path (falsifiable)" do
    m = start_model(15)
    generate = fn req -> model_generate(m, req) end

    rt = Runtime.new(%{in_process: generate, max_concurrent_turns: 1, registry: @registry})

    1..5
    |> Task.async_stream(
      fn i ->
        {:ok, w} = Runtime.spawn_agent(rt, Runtime.root(rt), "worker")
        Runtime.wake(rt, w, "job #{i}")
        r = Runtime.wait(rt, w, 5_000)
        assert r.status == "done"
      end,
      max_concurrency: 5,
      timeout: 10_000
    )
    |> Enum.each(fn {:ok, _} -> :ok end)

    stats = model_stats(m)

    assert stats.overlaps == 0,
           "gate FALSIFIED: #{stats.overlaps} call(s) observed >1 in flight (max_seen=#{stats.max_seen}) " <>
             "with max_concurrent_turns: 1 — the global turn gate did not wrap the :in_process transport"

    assert Runtime.max_observed_concurrent_turns(rt) == 1

    Runtime.shutdown(rt)
  end

  test "negative control: with max_concurrent_turns 5 the scripted model DOES observe overlap (proves the detector is real)" do
    m = start_model(15)
    generate = fn req -> model_generate(m, req) end

    rt = Runtime.new(%{in_process: generate, max_concurrent_turns: 5, registry: @registry})

    1..5
    |> Task.async_stream(
      fn i ->
        {:ok, w} = Runtime.spawn_agent(rt, Runtime.root(rt), "worker")
        Runtime.wake(rt, w, "job #{i}")
        Runtime.wait(rt, w, 5_000)
      end,
      max_concurrency: 5,
      timeout: 10_000
    )
    |> Enum.each(fn {:ok, _} -> :ok end)

    stats = model_stats(m)

    assert stats.overlaps > 0,
           "control INVALID: expected overlap with max_concurrent_turns: 5 and a 15ms hold, " <>
             "got 0 overlaps (max_seen=#{stats.max_seen}) — the detector proves nothing"

    Runtime.shutdown(rt)
  end
end
